cuML/src/decisiontree/kernels/col_condenser.cuh

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <thrust/sort.h>
+#include "atomic_minmax.h"
+
+template<typename T>
+__global__ void get_sampled_column_kernel(const T* __restrict__ column, T *outcolumn, const unsigned int* __restrict__ rowids, const int N) {
+
+	int tid = threadIdx.x + blockIdx.x * blockDim.x;
+	if (tid < N) {
+		int index = rowids[tid];
+		outcolumn[tid] = column[index];
+	}
+	return;
+}
+
+void get_sampled_labels(const int *labels, int *outlabels, unsigned int* rowids, const int n_sampled_rows, const cudaStream_t stream) {
+	int threads = 128;
+	get_sampled_column_kernel<int><<<(int)(n_sampled_rows / threads) + 1, threads, 0, stream>>>(labels, outlabels, rowids, n_sampled_rows);
+	CUDA_CHECK(cudaStreamSynchronize(stream));
```

Comment:
Is this sync needed here?

Reply:
Technically no in the current code version. The gini call that follows  launches the gini_kernel in the same stream as the get_sampled_column_kernel and there's no dependence on the CPU side code prior to that kernel launch. I can remove it.